### PR TITLE
Support version and timestamp for DeltaSharingClient.getMetadata

### PR DESCRIPTION
**Pull request by linzhou-db**
Support version and timestamp for DeltaSharingClient.getMetadata, added new unit tests.

Please only review changes under client/ repo, server changes should be covered by https://github.com/delta-io/delta-sharing/pull/376

### CODE REVIEW

**Review by charlenelyu-db — APPROVED** [review]

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala [hunk 1]
```diff
@@ -545,6 +554,16 @@ class DeltaSharingRestClient(
...
   }
 
+  private def getEncodedMetadataParams(
+      versionAsOf: Option[Long], timestampAsOf: Option[String]): String = {
+    val paramMap = versionAsOf.map("version" -> _.toString).toMap ++
+      timestampAsOf.map("timestamp" -> _).toMap
+    val params = paramMap.map {
+      case (key, value) => s"$key=${URLEncoder.encode(value)}"
+    }.mkString("&")
+    Option(params).map{x => if (x.nonEmpty) { "?" + x } else { "" }}.get
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala]: updated.

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala [hunk 2]
```diff
@@ -545,6 +554,20 @@ class DeltaSharingRestClient(
     }
   }
 
+  private def getEncodedMetadataParams(
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala:557]: Consider documenting this function.
An example encoded string will be useful to visualize what is happening.

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala [hunk 3]
```diff
@@ -209,12 +212,18 @@ class DeltaSharingRestClient(
...
+      versionAsOf: Option[Long] = None,
+      timestampAsOf: Option[String] = None): DeltaTableMetadata = {
     val encodedShareName = URLEncoder.encode(table.share, "UTF-8")
     val encodedSchemaName = URLEncoder.encode(table.schema, "UTF-8")
     val encodedTableName = URLEncoder.encode(table.name, "UTF-8")
+    val encodedParams = getEncodedMetadataParams(versionAsOf, timestampAsOf)
+
     val target = getTargetUrl(
-      s"/shares/$encodedShareName/schemas/$encodedSchemaName/tables/$encodedTableName/metadata")
+      s"/shares/$encodedShareName/schemas/$encodedSchemaName/tables/$encodedTableName/metadata" +
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala:225]: Do we need to update our protocol.md file for this change?

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala [hunk 4]
```diff
@@ -209,12 +212,18 @@ class DeltaSharingRestClient(
...
+        s"$encodedParams")
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala:226]: What if this request is sent to the server that does not support these new params?
Do we need to worry about compatibility issues for this?
Consider adding a comment.

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala (hunk 2 above)
Comment (linzhou-db) [reply client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala:557]: added.

**Pull request merged**